### ISSUE TITLE
setting defaults() for nested optional object attribute within parent optional object attribute using experimental `module_variable_optional_attrs` causes panic

### ISSUE THREAD

**Issue by marshall7m** [hashicorp/terraform#28046]
### Terraform Version

```
v0.15.0-beta1
```

### Terraform Configuration Files

```terraform
terraform {
  required_version = "0.15.0"
  experiments = [module_variable_optional_attrs]
}

variable "foo" {
  type =object({
    bar_one = optional(string)
    bar_two = optional(string)
    bar_three = optional(object({
      attr_one = string
      attr_two = optional(string)
    }))
  })
  default = {}
}

locals {
    y = defaults(var.foo, {bar_three = {
      attr_two = "default_two"}
    })
}

output "y" {
    value = local.y
}
```

### Debug Output

[Gist](https://gist.github.com/marshall7m/ef14d919cae7a361ac143b111a2c3566)
### Expected Behavior

if var.foo is not defined:

```
Changes to Outputs:
  + y = {
      + bar_one   = null
      + bar_three = null
      + bar_two   = null
    }
```

if var.foo.bar_three is not defined and other attributes are:

```
Changes to Outputs:
  + y = {
    bar_one = "hello"
    bar_two = "world"
  }
```

if var.foo.bar_three.attr_two is not defined but attr_one is:

```
Changes to Outputs:
  + y = {
    bar_three = {
      attr_one = "bar"
      attr_two = "test_two"
  }
```

if var.foo.bar_three.attr_one is not defined but attr_two is OR is not:

```
│ Error: Invalid default value for variable
│ 
│   on main.tf line 11, in variable "foo":
│   11:   default = {}
│ 
│ This default value is not compatible with the variable's type constraint: attribute "attr_one" is
│ required.
```

### Actual Behavior

See `Debug Output`
### Steps to Reproduce

1. Copy and paste code from `Terraform Configuration Files`
1. `terraform init`
3. `terraform apply`

**Comment by marshall7m** [comment]
Update:
Silly me, just realized that the version of Terraform I was using was `v0.15.0-beta1` and the fixes for this error are likely to be fixed in the upcoming `0.15.0-beta2`.  Will close for now but possibly reopen if `0.15.0-beta2` doesn't fix this error.

**Comment by alisdair** [comment]
Hi @marshall7m, thanks for reporting this. I happened to see the issue before you closed it and it turns out that this is not covered by the fixes already merged, so I'm reopening and taking a look.

In reading your report I was honestly a bit confused about what `defaults` _should_ do here, and it was only after working on a fix that I've arrived at a conclusion. Even now it seems a bit counter-intuitive to me. If you can share more about your real use case, I'd be interested to hear it!

There is one last thing to note about the expected behaviour you mentioned. If `var.foo` is not defined, you expected the result to be:

```
Changes to Outputs:
  + y = {}
```

However the optional attributes will be replaced with nulls, so the output will instead be:

```
Changes to Outputs:
  + y = {
      + bar_one   = null
      + bar_three = null
      + bar_two   = null
    }
```

**Comment by marshall7m** [comment]
> Hi @marshall7m, thanks for reporting this. I happened to see the issue before you closed it and it turns out that this is not covered by the fixes already merged, so I'm reopening and taking a look.
> 
> In reading your report I was honestly a bit confused about what `defaults` _should_ do here, and it was only after working on a fix that I've arrived at a conclusion. Even now it seems a bit counter-intuitive to me. If you can share more about your real use case, I'd be interested to hear it!
> 
> There is one last thing to note about the expected behaviour you mentioned. If `var.foo` is not defined, you expected the result to be:
> 
> ```
> Changes to Outputs:
>   + y = {}
> ```
> 
> However the optional attributes will be replaced with nulls, so the output will instead be:
> 
> ```
> Changes to Outputs:
>   + y = {
>       + bar_one   = null
>       + bar_three = null
>       + bar_two   = null
>     }
> ```

Hi @alisdair , thanks for taking a closer look than I did! Ah yes, the expected output should contain null attributes. I updated the issue. I also realized that I accidentally added a default for `attr_one` which may have made my report confusing.

In terms of my use case, the variable is for setting the environment configuration for an AWS Codebuild project.  The variable is defined as:

```
variable "codebuild_environment" {
  description = "Codebuild environment configuration"
  type = object({
    compute_type                = optional(string)
    image                       = optional(string)
    type                        = optional(string)
    image_pull_credentials_type = optional(string)
    environment_variables = optional(list(object({
      name  = string
      value = string
      type  = optional(string)
    })))
    privileged_mode = optional(bool)
    certificate     = optional(string)
    registry_credential = optional(object({
      credential          = optional(string)
      credential_provider = optional(string)
    }))
  })
  default = {}
}

locals {
  codebuild_environment = defaults(var.codebuild_environment, {
    compute_type = "BUILD_GENERAL1_SMALL"
    type = "LINUX_CONTAINER"
    image = "aws/codebuild/standard:3.0"
  })
}
```

As you can see above, the `environment_variables` attribute's type constraint should allow the user to optionally define a list of environment variables. Although, if they do define an environment variable object, the `name` and `value` attributes are required while the `type` attribute should be optional given that the AWS provider's `aws_codebuild_project` `environment_variable` `type` attribute is optional.

**Comment by alisdair** [comment]
Thanks for the extra detail! That's very helpful.

This configuration makes sense to me and seems like a nice use of this feature. I wanted to note that I don't think that `environment_variables = {}` in your defaults is doing anything, since `defaults` has nothing to copy from if it's an empty object. However, in verifying that thought locally I found another panic, which I'll have a fix for next week, so thanks for that too!

**Comment by marshall7m** [comment]
> Thanks for the extra detail! That's very helpful.
> 
> This configuration makes sense to me and seems like a nice use of this feature. I wanted to note that I don't think that `environment_variables = {}` in your defaults is doing anything, since `defaults` has nothing to copy from if it's an empty object. However, in verifying that thought locally I found another panic, which I'll have a fix for next week, so thanks for that too!

No problem!

Thanks for the tip and for fixing this problem in your PR. Excited to see it in action in `beta2`!

**Comment by korinne** [comment]
Hi all, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values.

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!
